File: android/circlebox-cloud/src/main/kotlin/com/circlebox/cloud/CircleBoxCloudConfig.kt
package com.circlebox.cloud

data class CircleBoxCloudConfig(
    val endpoint: String,
    val ingestKey: String,
    val region: String = "auto",
    val enableFragmentSync: Boolean = true,
    val flushIntervalSec: Long = 15,
    val maxQueueMB: Int = 20,
    val wifiOnly: Boolean = false,
    val retryMaxBackoffSec: Long = 900,
    val enableAutoFlush: Boolean = true,
    val autoExportPendingOnStart: Boolean = true,
    val immediateFlushOnHighSignal: Boolean = true,
    val enableUsageBeacon: Boolean = false,
    val usageBeaconKey: String? = null,
    val usageBeaconEndpoint: String? = null,
    val usageBeaconMode: CircleBoxCloudUsageMode = CircleBoxCloudUsageMode.CORE_CLOUD,
    val usageBeaconMinIntervalSec: Long = 300
) {
    init {
        require(endpoint.startsWith("http")) { "endpoint must be a valid http(s) URL" }
        require(ingestKey.startsWith("cb_live_")) { "ingestKey must use cb_live_ prefix" }
        require(flushIntervalSec >= 10) { "flushIntervalSec must be >= 10" }
        require(maxQueueMB > 0) { "maxQueueMB must be > 0" }
        require(retryMaxBackoffSec >= 30) { "retryMaxBackoffSec must be >= 30" }
        require(usageBeaconMinIntervalSec >= 30) { "usageBeaconMinIntervalSec must be >= 30" }
        if (!usageBeaconKey.isNullOrBlank()) {
            require(usageBeaconKey.startsWith("cb_usage_")) { "usageBeaconKey must use cb_usage_ prefix" }
        }
        if (!usageBeaconEndpoint.isNullOrBlank()) {
            require(usageBeaconEndpoint.startsWith("http")) { "usageBeaconEndpoint must be a valid http(s) URL" }
        }
    }
}

File: flutter/circlebox_cloud_flutter/lib/src/circlebox_cloud_config.dart
enum CircleBoxCloudUsageMode {
  offlineOnly('offline_only'),
  coreCloud('core_cloud'),
  coreAdapters('core_adapters'),
  coreCloudAdapters('core_cloud_adapters'),
  selfHost('self_host');

  const CircleBoxCloudUsageMode(this.wireValue);
  final String wireValue;
}

class CircleBoxCloudConfig {
  const CircleBoxCloudConfig({
    required this.endpoint,
    required this.ingestKey,
    this.region = 'auto',
    this.enableFragmentSync = true,
    this.flushIntervalSec = 15,
    this.maxQueueMb = 20,
    this.wifiOnly = false,
    this.retryMaxBackoffSec = 900,
    this.enableAutoFlush = true,
    this.autoExportPendingOnStart = true,
    this.immediateFlushOnHighSignal = true,
    this.enableUsageBeacon = false,
    this.usageBeaconKey,
    this.usageBeaconEndpoint,
    this.usageBeaconMode = CircleBoxCloudUsageMode.coreCloud,
    this.usageBeaconMinIntervalSec = 300,
  }) : assert(
         flushIntervalSec >= 10,
         'flushIntervalSec must be >= 10',
       ),
       assert(maxQueueMb > 0, 'maxQueueMb must be > 0'),
       assert(
         retryMaxBackoffSec >= 30,
         'retryMaxBackoffSec must be >= 30',
       ),
       assert(
         usageBeaconMinIntervalSec >= 30,
         'usageBeaconMinIntervalSec must be >= 30',
       );

  final Uri endpoint;
  final String ingestKey;
  final String region;
  final bool enableFragmentSync;
  final int flushIntervalSec;
  final int maxQueueMb;
  final bool wifiOnly;
  final int retryMaxBackoffSec;
  final bool enableAutoFlush;
  final bool autoExportPendingOnStart;
  final bool immediateFlushOnHighSignal;
  final bool enableUsageBeacon;
  final String? usageBeaconKey;
  final Uri? usageBeaconEndpoint;
  final CircleBoxCloudUsageMode usageBeaconMode;
  final int usageBeaconMinIntervalSec;
}
